derp/derp_client.go

```diff
@@ -520,6 +539,9 @@ func (c *Client) recvTimeout(timeout time.Duration) (m ReceivedMessage, err erro
 			if err != nil {
 				return nil, fmt.Errorf("invalid server info frame: %v", err)
 			}
+			if c.isWatcher && si.SupportsAtomicWatcher != true {
+				return nil, errors.New("incompatible mesh peer does not support new watch connections format; please upgrade")
```

Comment:
Question: instead of erroring out here, should we not just log that we cannot proceed meshing with this peer and proceed with rest of the derp operation? I'm thinking about the upgrade process- are we going to modify `--mesh-with` flag values? Else we probably won't be able to upgrade one by one if each derp errors out when it sees an older peer that does not support the new format

Reply:
Thanks, that's a better error for sure. The upgrade question is a good point and I want to test this upgrade in a test derp mesh and possible on a live derp region before merging. But I do think that this will end up looping in mesh_client.go here until the mesh peer is also upgraded:

        for ctx.Err() == nil {                                                                                                       
                _, lastConnGen, err = c.connect(c.newContext(), "derphttp.mesh_client.RunWatchConnectionLoop")                       
                if err != nil {                                                                                                      
                        logf("RunWatchConnectionLoop: initial connect() failed, retrying: %s", err)                                  
                        sleep(retryInterval)                                                                                         
                        continue                                                                                                     
                }

Reply:
My bad, because the server info comes in a separate packet after the connection is established, the loop will actually be here:

                      m, connGen, err := c.RecvDetail()                                                                            
                        if err != nil {                                                                                              
                                clear()                                                                                              
                                logf("RunWatchConnectionChanges: %v", err)                                                           
                                sleep(retryInterval)                                                                                 
                                break                                                                                                
                        }

But why guess when we can test. :)

Reply:
Thanks for the explanation, I followed the code up the stack- so when we upgrade a derp X, so if I understand correctly:

- as a client X will attempt to mesh with each of the peers concurrently and each of the attempts to mesh with non-upgraded peers will fail, remove the connection, sleep and retry till the peer has been upgraded. So after upgade it will be unaware of each of its older peers till the peer has been upgaded + the sleep period (5 seconds), but that should not block it meshing with upgaded peers.
And I guess the old peer would drop any forward packets that it would receive from X- except that X won't send it any.

- as a server, its http handler will fail for each peer client conection request that uses the old frameWatchConns mechanism (but this should not block it accepting connections from upgraded peers/other clients)

So, I guess I'm quite certain that the individual derps should keep functioning during upgrade and will auto-discover the peers. My only questions remains about what will happen to the forwarding mechanism during the upgrade? It looks like that any packets that would otherwise be forwarded between two derps in a region will get dropped if they aren't meshed because of version mismatch- is that potentially an issue?